snuba/migrations/snuba_migrations/replays/0001_replays.py

```diff
@@ -0,0 +1,96 @@
+    UInt,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+
+raw_columns: Sequence[Column[Modifiers]] = [
+    Column("replay_id", UUID()),
+    Column("sequence_id", UInt(16)),
+    Column("trace_ids", Array(UUID())),
+    Column("title", String()),
+    ### columns used by other sentry events
+    Column("project_id", UInt(64)),
+    # time columns
+    Column("timestamp", DateTime()),
+    # release/environment info
+    Column("platform", String(Modifiers(low_cardinality=True))),
+    Column("environment", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("release", String(Modifiers(nullable=True))),
+    Column("dist", String(Modifiers(nullable=True))),
+    Column("ip_address_v4", IPv4(Modifiers(nullable=True))),
+    Column("ip_address_v6", IPv6(Modifiers(nullable=True))),
+    # user columns
+    Column("user", String()),
+    Column("user_hash", UInt(64)),
+    Column("user_id", String(Modifiers(nullable=True))),
+    Column("user_name", String(Modifiers(nullable=True))),
+    Column("user_email", String(Modifiers(nullable=True))),
+    # sdk info
+    Column("sdk_name", String()),
+    Column("sdk_version", String()),
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    # internal data
+    Column("retention_days", UInt(16)),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.REPLAYS,
+                table_name="replays_local",
+                columns=raw_columns,
+                engine=table_engines.ReplacingMergeTree(
+                    storage_set=StorageSetKey.REPLAYS,
+                    order_by="(project_id, toStartOfDay(timestamp), cityHash64(replay_id), sequence_id)",
+                    partition_by="(retention_days, toMonday(timestamp))",
+                    settings={"index_granularity": "8192"},
+                    ttl="timestamp + toIntervalDay(retention_days)",
+                ),
+            ),
+        ]
+
+    def backwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.REPLAYS,
+                table_name="replays_local",
+            ),
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.REPLAYS,
+                table_name="replays_dist",
+                columns=raw_columns,
+                engine=table_engines.Distributed(
+                    local_table_name="replays_local",
+                    sharding_key="project_id",
```

Comment:
What is the reason for sharding the data by `project_id` versus sharding randomly? One disadvantage I can see with sharding by `project_id is that if there is a big project which uses replays a lot, the shards could become imbalanced.

Reply:
good point, I think i just chose this arbitrarily. I'll shard by `replay_id` instead.